src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests4.cs

```diff
@@ -0,0 +1,663 @@
+                //         Console.WriteLine(t is (3, 0, 5)); // false
+                Diagnostic(ErrorCode.ERR_BadArgCount, "(3, 0, 5)").WithArguments("Deconstruct", "3").WithLocation(12, 32),
+                // (12,32): error CS8129: No suitable Deconstruct instance or extension method was found for type 'C', with 3 out parameters and a void return type.
+                //         Console.WriteLine(t is (3, 0, 5)); // false
+                Diagnostic(ErrorCode.ERR_MissingDeconstruct, "(3, 0, 5)").WithArguments("C", "3").WithLocation(12, 32),
+                // (13,32): error CS1501: No overload for method 'Deconstruct' takes 4 arguments
+                //         Console.WriteLine(t is (3, 4, 5, 6)); // false
+                Diagnostic(ErrorCode.ERR_BadArgCount, "(3, 4, 5, 6)").WithArguments("Deconstruct", "4").WithLocation(13, 32),
+                // (13,32): error CS8129: No suitable Deconstruct instance or extension method was found for type 'C', with 4 out parameters and a void return type.
+                //         Console.WriteLine(t is (3, 4, 5, 6)); // false
+                Diagnostic(ErrorCode.ERR_MissingDeconstruct, "(3, 4, 5, 6)").WithArguments("C", "4").WithLocation(13, 32)
+                );
+        }
+
+        [Fact]
+        public void ITuple_05()
+        {
+            // - should match when input type extends ITuple and has no Deconstruct (type parameter)
+            var source =
+@"using System;
+using System.Runtime.CompilerServices;
+public class C : ITuple
+{
+    int ITuple.Length => 3;
+    object ITuple.this[int i] => i + 3;
+    public static void Main()
+    {
+        M(new C());
+    }
+    public static void M<T>(T t) where T: C
+    {
+        Console.WriteLine(t is (3, 4)); // false
+        Console.WriteLine(t is (3, 4, 5)); // TRUE
+        Console.WriteLine(t is (3, 0, 5)); // false
+        Console.WriteLine(t is (3, 4, 5, 6)); // false
+    }
+}
+";
+            var compilation = CreatePatternCompilation(source);
+            compilation.VerifyDiagnostics();
+            var expectedOutput =
+@"False
+True
+False
+False";
+            var compVerifier = CompileAndVerify(compilation, expectedOutput: expectedOutput);
+        }
+
+        [Fact]
+        public void ITuple_06()
+        {
+            // - should not match when input type extends ITuple and has Deconstruct (type parameter)
+            var source =
+@"using System;
+using System.Runtime.CompilerServices;
+public class C : ITuple
+{
+    int ITuple.Length => 3;
+    object ITuple.this[int i] => i + 3;
+    public static void Main()
+    {
+        M(new C());
+    }
+    public static void M<T>(T t) where T: C
+    {
+        Console.WriteLine(t is (3, 4)); // false
+        Console.WriteLine(t is (3, 4, 5)); // TRUE
+        Console.WriteLine(t is (3, 0, 5)); // false
+        Console.WriteLine(t is (3, 4, 5, 6)); // false
+    }
+    public void Deconstruct() {}
```

Comment:
Could you add a variant of such scenario (where the existence of `Deconstruct` interferes with binding to `ITuple`) but the `Deconstruct` method has a non-void return type?